### DISCUSSION TITLE
Benchmarking

### DISCUSSION THREAD

**Discussion by sorelfitzgibbon** [uclahs-cds/pipeline-generate-SQC-BAM#73] [General]
I'm collecting various benchmarking threads here.

1. [`mosdepth` and `CollectWgsMetrics`](https://github.com/uclahs-cds/pipeline-generate-SQC-BAM/pull/70#issuecomment-2237704226)

**Comment by sorelfitzgibbon** [comment]
@tyamaguchi-ucla @raagagrawal @yashpatel6 @pboutros These results have now been updated to include Raag's high coverage sample.  I think we should drop per-base-coverage output.  Let me know if you want plots.

## `mosdepth` vs `SAMtools depth`
The results below show that the two tools can be made to give almost identical per base results by setting `min-BQ 0`.  Only `SAMtools depth` has the option to change min-BQ, and using this does have a noticeable impact on per base coverages.  
Per base coverages are used for downstream statistical calculations requiring knowledge of how many sites could possibly have had a variant.  The differences when setting `SAMtools depth` to `min-BQ 20` remain small overall and could be acceptable, however I suggest **we should only output coverage windows with this pipeline and we should instead use the gVCFs from the `call-gSNP` pipeline for the non-variant reference quality information**, for these reasons:  
  - the g.vcfs are already being created by the pipeline
      - **avoid creating extra ~8 GB per-base output file**
  - the reference confidence scores within the g.vcfs are based on more than just base quality and they are **calculated after local reassembly** by `HaplotypeCaller` or `Mutect2`, thus presumably giving more accurate results.
  - `mosdepth` is much less resource intense: `3G` vs `26G` and parallelizable.

### Samples 
<img width="623" alt="image" src="https://github.com/user-attachments/assets/042b8ff1-8e8c-4895-924d-3d20f5911796">

### Resources
<img width="629" alt="image" src="https://github.com/user-attachments/assets/d85838af-a9e3-4165-aae8-d8f6ca9ed897">

###  Per base coverage differences
- parameters used   
    - `mosdepth --mapq 20`
    - `samtools depth -s -min-MQ 20 -min-BQ $x`
    - all other parameter options were default
    - more parameter option details [here](https://uclahs.app.box.com/integrations/officeonline/openOfficeOnline?fileId=1574735834153&activeCell=%27defaults)

#### Results for HG001 (30x), SARC (30x) and BAPR (80x) and min-BQ 0, 5, 10, 20.  
<img width="923" alt="image" src="https://github.com/user-attachments/assets/1e3ccf8f-4063-4d1a-a7d6-c7a26c2f1986">

#### Cause of non-base-quality differences
For the HG001 sample, of the 3,532 positions with coverage greater than `5x`, only 379 have coverage less than `100x` indicating the differences occur mostly at very high coverage sites (3/4 are > `1000x`)
    - the normal coverage sites have minor/rare issues, for example:
   
For the positions shown below with the 2 base deletion, `mosdepth` calls `26x` and `samtools` calls `20x`.  There are 6 shaded deletions which represent read pairs where one read has the deletion and the other doesn't.  (This is caused by one of the reads ending by the simple repeat region and aligning with a single mismatch instead of the 2 base deletion). `mosdepth
![image](https://github.com/user-attachments/assets/de83dabc-cfeb-44d2-a3d1-f535c6b42e83)

**Reply by yashpatel6** [reply]
@tyamaguchi-ucla @raagagrawal Any thoughts on this discussion thread from before regarding coverage calculation tools?

**Reply by raagagrawal** [reply]
I don't have much contribution beyond what Sorel has said. I think mosdepth offers a lot of performance improvements at the cost of not having Base Quality minimums. I assume a sample with systematically low base quality would then show big differences between samtools and mosdepth - but that this would be obvious from the standard QC we run in tools-QC-scripts.
